### PR TITLE
Make signal delegate static to avoid memory allocation

### PR DESCRIPTION
**Pull request by CedNaru**
Implement #694 

I mostly modified the Signal generation so it can create its own delegate. I used the opportunity to split generation into smaller functions. If you want to see the truly new stuff, look at the `SignalGenerationService::generateSignalDelegateCompanion`.

The api generation for signals went from this:
```kt
  public val childEnteredTree: Signal1<Node> by signal("node")
```
to 
```kt
  public val childEnteredTree: Signal1<Node> by Signal1
```

The reason for the change is that the previous method was actually creating a field in the class for each signal (as explained in the linked proposal), you can only remove that field if the delegation is done on something known at compile time. This sadly doesn't include inline functions. So even if I transform the `signal()` function to
```kt
  inline fun <P0> signal1(p0: String) = Signal1
```
It won't work (I tried, and the bytecode generated still has fields for delegates, even if they all point to the same object).

It doesn't matter that the parameter strings are no longer generated, they were useless anyway. Their only function is to name the user created signals in the entry gen so it can be safely be removed in the api gen.

Note: 
I am not fully satisfied. It makes wrappers much lighter (for example, for Node2D, we will save more than 150 Bytes per instance) but this only applies to wrappers. I didn't find a way to make it work for scripts yet.
The main issue is user created signals need string information (for naming the parameters) from the signal() method, which means we can't directly delegate to the static object, which means creating a field.

The only way around it, that I can imagine so far, is to change the way user signals are written. (NOT IMPLEMENTED, just discussing the idea)
From
```kt
 @RegisterSignal
 val mySignal: Signal1<Int> by signal("number")
```
to 
```kt
 @RegisterSignal("number")
 val mySignal: Signal1<Int> by Signal1
```
The main issue with this approach is that we need a `vararg string` parameter in the annotation, which doesn't provide any safety regarding the number of require parameters to name. It's error-prone, as we can't rely on a method signature directly.
On top of that, it goes against our plan to remove the annotation for signals as proposed in #606

### CODE REVIEW

**Review by chippmann — APPROVED** [review]

**Comment by chippmann** [comment]
You proposal for user scripts is IMO not an option. IMO safety is more important for user scripts than saving a few bytes in this case.
For the wrappers though, this PR is perfect

**Comment by CedNaru** [comment]
Well there is a way to keep safety even for script signals. We just need a @RegisterSignal0, @RegisterSignal1, but it's just cumbersome to use and, again, against what we want to do.
I also was never satisfied about how we use this `signal(string)` method, the string is pure metadata that have no consequence on the code itself, the parameters are not even used by the method, they are only there for the sake of being read by KSP. 
So even the current state is quite dirty imo. But I don't have any idea that would fix it and not be a pain to use.

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/SignalGenerationService.kt [hunk 1]
```diff
@@ -23,256 +24,285 @@ import godot.tools.common.constants.GODOT_CALLABLE
 import godot.tools.common.constants.GODOT_ERROR
 import godot.tools.common.constants.GODOT_OBJECT
 import godot.tools.common.constants.godotCorePackage
+import godot.tools.common.constants.kotlinReflectPackage
 import kotlin.reflect.KCallable
 
 class SignalGenerationService : ISignalGenerationService {
+
+    // Helper that provide utilities to handle generic classes.
+    private class GenericClassNameInfo(
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/services/impl/SignalGenerationService.kt:33]: Could be cool to have a toTypeSpec method in this class that does:
```kotlin
TypeSpec
            .classBuilder(className)
            .addTypeVariables(genericTypes)
```

File: kt/api-generator/src/main/kotlin/godot/codegen/services/impl/SignalGenerationService.kt [hunk 2]
```diff
@@ -23,256 +24,285 @@ import godot.tools.common.constants.GODOT_CALLABLE
...
+            .addTypeVariables(genericClassNameInfo.genericTypes)
+            .addParameters(
+                (0..<argCount)
+                    .map {
+                        ParameterSpec.builder("p$it", STRING).build()
+                    }
+            )
+            .addCode(
+                CodeBlock.of(
+                    "return %T<%T, %T> { thisRef, property ->  %T(thisRef, property.name)}",
```
Comment (piiertho) [inline kt/api-generator/src/main/kotlin/godot/codegen/services/impl/SignalGenerationService.kt:256]: replace spaces with `·`

**Review by piiertho — CHANGES_REQUESTED** [review]

File: kt/tools-common/src/main/kotlin/godot/tools/common/constants/Packages.kt [hunk 3]
```diff
@@ -17,6 +17,7 @@ const val variantCasterPackage = "$godotCorePackage.${GodotKotlinJvmTypes.varian
 
 const val kotlinPackage = "kotlin"
 const val kotlinCollectionsPackage = "$kotlinPackage.collections"
+const val kotlinReflectPackage = "$kotlinPackage.reflect"
 const val kotlinTextPackage = "$kotlinPackage.text"
-const val kotlinCoroutinePackage = "kotlin.coroutines"
+const val kotlinCoroutinePackage = "$kotlinPackage..coroutines"
```
Comment (piiertho) [inline kt/tools-common/src/main/kotlin/godot/tools/common/constants/Packages.kt:22]: Why two dots here ?

**Review by piiertho — APPROVED** [review]

**Review by chippmann — APPROVED** [review]
Approving under the assumption that this will not be released as is.
As discussed on discord, I'm not fond of the reduction in ease of use of user script signal definitions. The requirement of an explicit type declaration is one thing, but the not very descriptive error message for users not familiar with the internal workings of our module is a completely separate topic.
While the memory reduction is important, ease of use is as well.

Hence on discord we agreed on a different signal definition implementation which will be implemented as a follow up PR.

**Comment by chippmann** [comment]
image
This is one example of such an error message.

From the error, it's not clear that what is missing is an explicit type declaration `: Signal1<Type>`.
Especially since the provided quick fix options by the IDE are misleading

File: kt/godot-library/src/main/kotlin/godot/gen/godot/core/Signals.kt [hunk 4]
```diff
@@ -75,14 +88,17 @@ public class Signal2<P0, P1>(
...
+  public companion object {
+    public operator fun <P0, P1> getValue(thisRef: Object, `property`: KProperty<*>):
+        Signal2<P0, P1> = Signal2(thisRef, property.name)
+  }
 }
 
 public inline fun <reified P0, reified P1> Signal2<P0, P1>.connect(flags: Int = 0, noinline
     method: (p0: P0, p1: P1) -> Unit): Error = connect(method.asCallable(), flags)
 
-@Suppress("UNUSED_PARAMETER")
```
[the remark this answers is not in the document]
Comment (CedNaru) [reply kt/godot-library/src/main/kotlin/godot/gen/godot/core/Signals.kt:83]: It's defined at the file level now.

**Pull request merged**